### PR TITLE
Fix Cyclic dependencies in experimental/cas/main branch when building with -DBUILD_SHARED_LIBS=ON

### PR DESCRIPTION
**Pull request by rastogishubham**

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/tools/lto/CMakeLists.txt [hunk 1]
```diff
@@ -12,6 +12,8 @@ set(LLVM_LINK_COMPONENTS
   MCDisassembler
   Support
   Target
+  CAS
+  RemoteNullService
```
Comment (cachemeifyoucan) [inline llvm/tools/lto/CMakeLists.txt]: This is almost always wrong because this target doesn't exist when you turn on `LLVM_CAS_ENABLE_REMOTE_CACHE`.

You probably want to use `RemoteCachingService`. Same for all other references.

**Review by cachemeifyoucan — CHANGES_REQUESTED** [review]
LGTM in general but you want to test other configuration.

To enable remote cache, you want to `brew install grpc` before turning on that variable.

File: clang/tools/clang-check/CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ set( LLVM_LINK_COMPONENTS
   ${LLVM_TARGETS_TO_BUILD}
   Option
   Support
+  MCCAS
```
Comment (adrian-prantl) [inline clang/tools/clang-check/CMakeLists.txt]: This looks suspicious because neither MC or CAS are listed. Why is this needed?

File: clang/unittests/Basic/CMakeLists.txt [hunk 3]
```diff
@@ -1,5 +1,6 @@
 set(LLVM_LINK_COMPONENTS
   Support
+  CAS
```
Comment (adrian-prantl) [inline clang/unittests/Basic/CMakeLists.txt:3]: What's the dependency here?

File: llvm/tools/llvm-lto/CMakeLists.txt [hunk 4]
```diff
@@ -15,6 +15,9 @@ set(LLVM_LINK_COMPONENTS
   Support
   Target
   TargetParser
+  CAS
+  RemoteCachingService
+  RemoteNullService
```
Comment (adrian-prantl) [inline llvm/tools/llvm-lto/CMakeLists.txt]: Just for my understanding: Why does this depend on CAS but not on MCCAS?

Comment (cachemeifyoucan) [reply llvm/tools/llvm-lto/CMakeLists.txt]: This is reasonable. We only have full object caching in ThinLTO.

File: clang/unittests/Basic/CMakeLists.txt (hunk 3 above)
Comment (rastogishubham) [reply clang/unittests/Basic/CMakeLists.txt:3]: Without it I get

`FAILED: tools/clang/unittests/Basic/BasicTests 
: && /usr/bin/clang++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -fno-common -Woverloaded-virtual -Wno-nested-anon-types -g -arch arm64 -isysroot /Users/shubham/apple-internal/Xcode-Summit/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/CharInfoTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/DarwinSDKInfoTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/DiagnosticTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/FileEntryTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/FileManagerTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/LineOffsetMappingTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/SanitizersTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/SarifTest.cpp.o tools/clang/unittests/Basic/CMakeFiles/BasicTests.dir/SourceManagerTest.cpp.o -o tools/clang/unittests/Basic/BasicTests  -Wl,-rpath,/Users/shubham/Development/llvm-project-cas/llvm-project/build_ninja/lib  lib/libllvm_gtest_main.dylib  lib/libclangAST.dylib  lib/libclangLex.dylib  lib/libLLVMTestingSupport.dylib  lib/libclangBasic.dylib  lib/libllvm_gtest.dylib  lib/libLLVMSupport.dylib && :
Undefined symbols for architecture arm64:
  "llvm::cas::ObjectStore::getProxy(llvm::cas::ObjectRef)", referenced from:
      (anonymous namespace)::FileManagerTest_CASProvider_Test::TestBody() in FileManagerTest.cpp.o
  "llvm::cas::createInMemoryCAS()", referenced from:
      (anonymous namespace)::FileManagerTest_CASProvider_Test::TestBody() in FileManagerTest.cpp.o
  "llvm::cas::createCASProvidingFileSystem(std::__1::shared_ptr<llvm::cas::ObjectStore>, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>)", referenced from:
      (anonymous namespace)::FileManagerTest_CASProvider_Test::TestBody() in FileManagerTest.cpp.o
ld: symbol(s) not found for architecture arm64`

File: clang/tools/clang-check/CMakeLists.txt (hunk 2 above)
Comment (rastogishubham) [reply clang/tools/clang-check/CMakeLists.txt]: Yep, I was able to remove that one

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/RemoteCachingService/CMakeLists.txt [hunk 5]
```diff
@@ -1,27 +1,32 @@
 if (LLVM_CAS_ENABLE_REMOTE_CACHE)
 add_subdirectory(RemoteCacheProto)
 add_subdirectory(RemoteCacheServer)
```
Comment (cachemeifyoucan) [inline llvm/lib/RemoteCachingService/CMakeLists.txt:3]: Any reason why you decide not to expand these two directories?

File: llvm/lib/RemoteCachingService/CMakeLists.txt [hunk 6]
```diff
@@ -1,27 +1,32 @@
...
-set(LLVM_LINK_COMPONENTS
-  RemoteNullService
-  )
-endif()
+  include_directories($<TARGET_PROPERTY:LLVMRemoteCacheKVProto,BINARY_DIR>)
+  include_directories($<TARGET_PROPERTY:LLVMRemoteCacheCASProto,BINARY_DIR>)
 
+  add_llvm_component_library(LLVMRemoteCachingService
+  RemoteCachingService.cpp
+  Client/Client.cpp
```
Comment (cachemeifyoucan) [inline llvm/lib/RemoteCachingService/CMakeLists.txt]: I suggest you only use one `add_llvm_component_library` and only set the sharing setting once.

You can declare different variable in the branch and set
```
add_llvm_component_library(LLVMRemoteCachingService
  $REMOTE_CACHE_SOURCE
  RemoteCachingService.cpp

  LINK_COMPONENTS
   Support
   CAS
)
```

File: llvm/lib/RemoteCachingService/CMakeLists.txt (hunk 5 above)
Comment (rastogishubham) [reply llvm/lib/RemoteCachingService/CMakeLists.txt:3]: I may be wrong but because the 

`generate_llvm_protos_library(LLVMRemoteCacheKVProto "compilation_caching_kv.proto" GRPC)
generate_llvm_protos_library(LLVMRemoteCacheCASProto "compilation_caching_cas.proto" GRPC)`

lines are in the `RemoteCacheProto` lib, the header `compilation_caching_cas.grpc.pb.h` is not generated during the compilation step for Client.cpp because the link step comes after, and so I get a build error like:

`/Users/shubham/Development/llvm-project-cas/llvm-project/llvm/lib/RemoteCachingService/RemoteCacheServer/RemoteCacheServer.cpp:12:10: fatal error: 'compilation_caching_cas.grpc.pb.h' file not found
`

I could probably flatten `RemoteCacheServer` in though, but I decided to only flatten the ones that were actually causing an issue

Comment (rastogishubham) [reply llvm/lib/RemoteCachingService/CMakeLists.txt:3]: Okay I was able to flatten RemoteCacheServer, but seems like I will not be able to flatten RemoteCacheProto

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/RemoteCachingService/CMakeLists.txt (hunk 5 above)
Comment (akyrtzi) [reply llvm/lib/RemoteCachingService/CMakeLists.txt:3]: Could we keep `RemoteCacheServer` files as a separate library? Only `llvm-remote-cache-test` tool depends on that, nothing else needs to link to that code.

Comment (cachemeifyoucan) [reply llvm/lib/RemoteCachingService/CMakeLists.txt:3]: Sure, we can keep `RemoteCacheServer` in its old structure.

Maybe it is cleaner if we fold NullService into Service and select implementation using macros but it is not bad in its current status.

File: llvm/lib/RemoteCachingService/CMakeLists.txt [hunk 7]
```diff
@@ -1,27 +1,26 @@
...
-add_subdirectory(NullService)
+)
 
-set(LLVM_LINK_COMPONENTS
-  RemoteNullService
-  )
+  include_directories($<TARGET_PROPERTY:LLVMRemoteCacheKVProto,BINARY_DIR>)
+  include_directories($<TARGET_PROPERTY:LLVMRemoteCacheCASProto,BINARY_DIR>)
+
+  set(REMOTE_CACHE_SOURCE "Client/Client.cpp" "CAS/GRPCRelayCAS.cpp" "RemoteCacheServer/LLVMCASCacheProvider.cpp" "RemoteCacheServer/RemoteCacheServer.cpp")
```
Comment (cachemeifyoucan) [inline llvm/lib/RemoteCachingService/CMakeLists.txt]: Can you match the indentation like other `set`?

Also not sure if quotation mark is needed.

Comment (rastogishubham) [reply llvm/lib/RemoteCachingService/CMakeLists.txt]: I thought clang-format would do it, but now that @akyrtzi wanted RemoteCacheServer to be it's own dylib, the set is only 66 characters long

**Comment by rastogishubham** [comment]
@swift-ci please test

File: llvm/lib/RemoteCachingService/CMakeLists.txt (hunk 5 above)
Comment (rastogishubham) [reply llvm/lib/RemoteCachingService/CMakeLists.txt:3]: @akyrtzi done!

File: llvm/lib/RemoteCachingService/CMakeLists.txt (hunk 7 above)
Comment (cachemeifyoucan) [reply llvm/lib/RemoteCachingService/CMakeLists.txt]: clang-format doesn't do CMake files. Usually it is formatted into one word on each line (see `set` above)
```
set(REMOTE_CACHE_SOURCE
  Client/Client.cpp
  CAS/GRPCReplayCAS.cpp
  ...
)
```
Something like this.

Comment (cachemeifyoucan) [reply llvm/lib/RemoteCachingService/CMakeLists.txt]: And the indentation should be the same for all the lines in `if()`

Comment (rastogishubham) [reply llvm/lib/RemoteCachingService/CMakeLists.txt]: Okay sounds good, I can fix that!

**Comment by rastogishubham** [comment]
@swift-ci please test

**Review by cachemeifyoucan — APPROVED** [review]
Thanks for handling this!

**Comment by rastogishubham** [comment]
Anything for the building blocks team!
